pkg/reconciler/controllers/syncer.go

```diff
@@ -0,0 +1,343 @@
+	controllerCtx context.Context
+
+	// doneCh indicates the Syncer stops processing new requests.
+	// It is a signal for the Finalizer to continue to finalize managed resources.
+	doneCh chan struct{}
+
+	// scope is the scope of the reconciler, either a namespace or ':root'
+	scope declared.Scope
+
+	// pollingPeriod is the period of time between checking the filesystem for
+	// source updates to sync.
+	pollingPeriod time.Duration
+
+	// ResyncPeriod is the period of time between forced re-sync from source
+	// (even without a new commit).
+	resyncPeriod time.Duration
+
+	// retryPeriod is how long the parser waits between retries, after an error.
+	retryPeriod time.Duration
+
+	// statusUpdatePeriod is how long the parser waits between updates of the
+	// sync status, to account for management conflict errors from the remediator.
+	statusUpdatePeriod time.Duration
+
+	reimportTimer     *time.Timer
+	resyncTimer       *time.Timer
+	retryTimer        *time.Timer
+	statusUpdateTimer *time.Timer
+
+	// parser is the actor that parses the source configs from file format to unstructured Kubernetes objects,
+	// and sends K8s API calls to set the status field. It also contains an updater
+	// that applies the declared resources to the cluster, and update watches for drift correction.
+	parser parse.Parser
+
+	// state stores the shared state across multiple Syncer reconciliations.
+	state *parse.ReconcilerState
+
+	// wg is the WaitGroup for the Syncer.
+	wg sync.WaitGroup
+
+	// lifecycleMux guards live status of the Syncer.
+	lifecycleMux sync.Mutex
+	// canSync indicates whether the Syncer can start processing requests.
+	canSync bool
+}
+
+// NewSyncer initializes the Syncer controller.
+func NewSyncer(controlCtx context.Context, scope declared.Scope, parser parse.Parser,
+	state *parse.ReconcilerState, pollingPeriod, resyncPeriod, retryPeriod,
+	statusUpdatePeriod time.Duration, doneCh chan struct{}) *Syncer {
+	return &Syncer{
+		controllerCtx:      controlCtx,
+		scope:              scope,
+		pollingPeriod:      pollingPeriod,
+		resyncPeriod:       resyncPeriod,
+		retryPeriod:        retryPeriod,
+		statusUpdatePeriod: statusUpdatePeriod,
+		reimportTimer:      time.NewTimer(pollingPeriod),
+		resyncTimer:        time.NewTimer(resyncPeriod),
+		retryTimer:         time.NewTimer(retryPeriod),
+		statusUpdateTimer:  time.NewTimer(statusUpdatePeriod),
+		parser:             parser,
+		state:              state,
+		doneCh:             doneCh,
+		canSync:            true, // The syncer should be able to sync at any time unless the controller context is canceled.
+	}
+}
+
+// Reconcile processes the Syncer request to apply the source configs to the cluster.
+func (s *Syncer) Reconcile(_ context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if errors.Is(s.controllerCtx.Err(), context.Canceled) {
+		// Drain the request queue
+		klog.V(3).Infof("Stop syncing new events as the context is canceled..")
```

Comment:
Suggest:
`klog.V(3).Info("Skip syncing because the Syncer is stopped")`

Reply:
Not sure how to modify that to include the request, but it seems like it should. I'm not really clear what requests trigger reconciles.